File: SymPy/mwe.py
import QSymPy as qp
import sympy as sp

qc = qp.QuantumCircuit(num_qubits=3)
qc.add_gate(name='H', qubits_t=[0], step=1)
qc.add_gate(name='CX', qubits_t=[1], qubits_c=[0], step=2)
qc.add_gate(name='CX', qubits_t=[2], qubits_c=[0], step=2)

qc.assemble_symbolic_unitary()

qs = qp.QuantumState(num_qubits=3)

qs2 = qc.unitary @ qs.state

print(qs2)

print(sp.diff(qs2, qc.gate_collection.collections['CX'][0].matrix[1, 1]))

File: SymJl/QCSymjl_Tests/mwe.jl
import Symbolics
import QCSym

#include("../QSymJl/QSymJl.jl")
#include("../QSym.jl/src/QSym.jl")

qc = QCSym.Circuits.QuantumCircuit(name="SuperDuperNiceCircuit")
println(qc)
q_reg_1 = QCSym.Circuits.add_qreg(qc, "q_reg_1", 2)
c_reg_1 = QCSym.Circuits.add_creg(qc, "c_reg_1", 2)
println(qc)

QCSym.Circuits.add_gate(qc, QCSym.Gates.H_Gate; qubits_t=[q_reg_1[1]], step=1, is_treat_numeric_only=false)
#println(qc)
 QCSym.Circuits.add_gate(qc, QCSym.Gates.H_Gate; name_prefix="H2", qubits_t=[q_reg_1[2]], step=1, is_treat_numeric_only=false)
# QCSym.Circuits.add_gate(qc, QCSym.Gates.H_Gate; name_prefix="H3", qubits_t=[q_reg_1[1]], step=2, is_treat_numeric_only=false)

QCSym.Circuits.add_gate(qc, QCSym.Gates.CX_Gate; qubits_t=[q_reg_1[2]], qubits_c=[q_reg_1[1]], step=2, is_treat_numeric_only=false)

QCSym.Circuits.add_gate(qc, QCSym.Gates.U_Gate; qubits_t=[q_reg_1[1]], step=3, is_treat_numeric_only=false, is_treat_alt_only=true)

unitary = QCSym.Circuits.assemble_symbolic_unitary(qc, true, false)
display(Symbolics.simplify(unitary))

#unitary2 = Symbolics.substitute.(unitary, (Dict(zip(qc.gatecollection.collections[QCSym.Gates.CX_Gate][1].atomics, qc.gatecollection.collections[QCSym.Gates.CX_Gate][1].matrix_numeric)),))
#display(unitary2)
#println(qc.gatecollection.collections[QCSym.Gates.CX_Gate])

